File: Null/Null.go
package main

import "fmt"

type AbstractCustomer interface {
	isNil() bool
	getName() string
}

type RealCustommer struct {
	name string
}

func(r *RealCustommer)isNil() bool{
	return false;
}

func(r *RealCustommer)getName() string{
	return r.name;
}

type NullCustommer struct {
	name string
}

func(r *NullCustommer)isNil() bool{
	return true;
}

func(r *NullCustommer)getName() string{
	return "not available in customer database";
}

type CustomerFactory struct {
	names []string
}

func(c *CustomerFactory)getCustomer(name string) AbstractCustomer{
	c.names=[]string{"Rob", "Job", "Juile"}
	for _,n:=range c.names{
		if(n==name){
			return &RealCustommer{n}
		}
	}
	return new(NullCustommer)
}

func main(){
	custFac:=new(CustomerFactory);
	customer1 := custFac.getCustomer("Rob");
	customer2 := custFac.getCustomer("Bob");
	customer3 := custFac.getCustomer("Juile");
	customer4 := custFac.getCustomer("Laura");
	fmt.Println("Customers:");
	fmt.Println(customer1.getName())
	fmt.Println(customer2.getName())
	fmt.Println(customer3.getName())
	fmt.Println(customer4.getName())
}

File: Null/Null.php
<?php

/**
 * Created by PhpStorm.
 * User: jamesyan
 * Date: 2017/3/25
 * Time: 16:15
 */
abstract class AbstractCustomer
{
    protected $name;

    public abstract function isNil();

    public abstract function getName();
}

class RealCustommer extends AbstractCustomer
{
    public function __construct($name)
    {
        $this->name = $name;
    }

    public function getName()
    {
        // TODO: Implement getName() method.
        return $this->name;
    }

    public function isNil()
    {
        // TODO: Implement isNil() method.
        return false;
    }
}

class NullCustommer extends AbstractCustomer
{
    public function getName()
    {
        // TODO: Implement getName() method.
        return "not available in customer database";
    }

    public function isNil()
    {
        // TODO: Implement isNil() method.
        return true;
    }
}


class CustomerFactory
{
    public static $names = array("Rob", "Job", "Juile");

    public static function getCustomer($name)
    {

        foreach(self::$names as $n){
            if($n==$name){
                return new RealCustommer($name);
            }
        }
        return new NullCustommer();
    }
}


//demo
$customer1=CustomerFactory::getCustomer("Rob");
$customer2=CustomerFactory::getCustomer("Bob");
$customer3=CustomerFactory::getCustomer("Juile");
$customer4=CustomerFactory::getCustomer("Laura");
var_dump("Customers:");
var_dump($customer1->getName());
var_dump($customer2->getName());
var_dump($customer3->getName());
var_dump($customer4->getName());